solana/programs/01_hello_world/src/message.rs

```diff
@@ -46,26 +46,32 @@ impl AnchorSerialize for HelloWorldMessage {
 }
 
 impl AnchorDeserialize for HelloWorldMessage {
-    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
-        match buf[0] {
-            PAYLOAD_ID_ALIVE => Ok(HelloWorldMessage::Alive {
-                program_id: Pubkey::try_from(&buf[1..33]).unwrap(),
-            }),
+    fn deserialize_reader<R: io::Read>(reader: &mut R) -> io::Result<Self> {
+        let mut variant = 0;
+        reader.read_exact(slice::from_mut(&mut variant))?;
+        match variant {
+            PAYLOAD_ID_ALIVE => {
+                let mut buf = [0; 32];
+                reader.read_exact(&mut buf)?;
+                Ok(HelloWorldMessage::Alive {
+                    program_id: Pubkey::try_from(buf).unwrap(),
+                })
```

Comment:
In fact, seeing how this read pattern repeats everywhere, we probably want this not just for uints but apparently also Pubkeys

Reply:
I'm using read now, as suggested below.